sessiontxn/staleread/processor.go

```diff
@@ -0,0 +1,223 @@
+type Processor interface {
+	// IsStaleness indicates that whether we should use the staleness
+	IsStaleness() bool
+	// GetStalenessInfoSchema returns the information schema if it is stale read, otherwise returns nil
+	GetStalenessInfoSchema() infoschema.InfoSchema
+	// GetStalenessReadTS returns the ts if it is stale read, otherwise returns 0
+	GetStalenessReadTS() uint64
+	// GetStalenessTSEvaluatorForPrepare returns a function that will be used by prepare to evaluate ts
+	GetStalenessTSEvaluatorForPrepare() StalenessTSEvaluator
+
+	// OnSelectTable will be called when process table in select statement
+	OnSelectTable(tn *ast.TableName) error
+}
+
+type baseProcessor struct {
+	sctx       sessionctx.Context
+	txnManager sessiontxn.TxnManager
+
+	evaluated   bool
+	ts          uint64
+	tsEvaluator StalenessTSEvaluator
+	is          infoschema.InfoSchema
+}
+
+func (p *baseProcessor) init(sctx sessionctx.Context) {
+	p.sctx = sctx
+	p.txnManager = sessiontxn.GetTxnManager(sctx)
+}
+
+func (p *baseProcessor) IsStaleness() bool {
+	return p.ts != 0
+}
+
+func (p *baseProcessor) GetStalenessInfoSchema() infoschema.InfoSchema {
+	return p.is
+}
+
+func (p *baseProcessor) GetStalenessReadTS() uint64 {
+	return p.ts
+}
+
+func (p *baseProcessor) GetStalenessTSEvaluatorForPrepare() StalenessTSEvaluator {
+	return p.tsEvaluator
+}
+
+func (p *baseProcessor) OnSelectTable(_ *ast.TableName) error {
+	return errors.New("not supported")
+}
+
+func (p *baseProcessor) setAsNonStaleRead() error {
+	return p.setEvaluatedValues(0, nil, nil)
+}
+
+func (p *baseProcessor) setEvaluatedTS(ts uint64) (err error) {
+	is, err := domain.GetDomain(p.sctx).GetSnapshotInfoSchema(ts)
+	if err != nil {
+		return err
+	}
+	is = temptable.AttachLocalTemporaryTableInfoSchema(p.sctx, is)
+	return p.setEvaluatedValues(ts, is, func(sctx sessionctx.Context) (uint64, error) {
+		return ts, nil
+	})
+}
+
+func (p *baseProcessor) setEvaluatedEvaluator(evaluator StalenessTSEvaluator) error {
+	ts, err := evaluator(p.sctx)
+	if err != nil {
+		return err
+	}
+
+	is, err := domain.GetDomain(p.sctx).GetSnapshotInfoSchema(ts)
+	is = temptable.AttachLocalTemporaryTableInfoSchema(p.sctx, is)
+	if err != nil {
+		return err
+	}
+
+	return p.setEvaluatedValues(ts, is, evaluator)
+}
+
+func (p *baseProcessor) setEvaluatedValues(ts uint64, is infoschema.InfoSchema, tsEvaluator StalenessTSEvaluator) error {
+	if p.evaluated {
+		return errors.New("already evaluated")
+	}
+
+	p.ts = ts
+	p.is = is
+	p.evaluated = true
+	p.tsEvaluator = tsEvaluator
+	return nil
+}
+
+type staleReadProcessor struct {
+	baseProcessor
+	stmtAsOfTs uint64
+}
+
+// NewStaleReadProcessor creates a new stale read processor
+func NewStaleReadProcessor(sctx sessionctx.Context) Processor {
+	p := &staleReadProcessor{}
+	p.init(sctx)
+	return p
+}
+
+// OnSelectTable will be called when process table in select statement
+func (p *staleReadProcessor) OnSelectTable(tn *ast.TableName) error {
```

Comment:
Why the comment for explaining the code logic removed?

Reply:
I'll complete it later

Reply:
done